Chimera/spm fetch dependencies

laxmorek opened the pull request:
Working on #1674
Part of #2394

### Short description 📝

This PR adds logic for fetching SPM dependencies defined in `Dependencies.swift`. 

### Implementation 👩‍💻👨‍💻

**The flow:**
1. Build `Package.swift` 
2. Call `swift package` to resolve dependencies 
3. Save `Package.resolved` and resolved dependencies (`.build` directory) in `/Tuist/Dependencies` directory. 

ad.1
Right now as a temporary solution I generate `Package.swift` as a `String` value. We will be able to upgrade it when Apple resolves [this](https://forums.swift.org/t/pitch-package-editor-commands/42224).

<img width="807" alt="Zrzut ekranu 2021-03-14 o 22 00 27" src="https://user-images.githubusercontent.com/4774319/111084015-c53cd000-8510-11eb-9481-bf37e9f7b7dd.png">
<img width="1210" alt="Zrzut ekranu 2021-03-14 o 22 00 50" src="https://user-images.githubusercontent.com/4774319/111084017-c66dfd00-8510-11eb-969b-eea8de3934ce.png">

### Questions ❓

I hardcoded `// swift-tools-version:5.3` value for `Package.swift`. Should we allow users to define `swift-tools-version` in `Dependencies.swift` manifest file? 

If the answer is yes then I can see two strategies:
1. Defining `swift-tools-version` using `ProjectDescription.Version` model and linting value provided by user. 
2. Creating new `enum` with predefined values.

```swift
enum SwiftToolsVersion {
    case v5_3
    case v5_4
    case v6_0
}
```

_Advantage_ of this solution is always correct value provided by user, no linting needed. 
_Disadvantage_ of this solution is maintenance work that we will have to do for every new version of `swift-tools-version`.

One way or another `Dependencies.swift` may look like below:

```swift
let dependencies = Dependencies(
    swiftPackageManager: .swiftPackageManager(
        [
            .remote(url: "https://github.com/Alamofire/Alamofire.git", requirement: .exact("5.2.0"))
        ],
        swiftToolsVersion: "5.4.0"
    )
)
```

pepibumur commented on Sources/TuistSupport/Constants.swift:
```diff
@@ -33,7 +33,9 @@ public enum Constants {
         public static let cartfileResolvedName = "Cartfile.resolved"
+        public static let packageResolvedName = "Package.resolved"
         public static let carthageDirectoryName = "Carthage"
+        public static let swiftPackageManagerDirectoryName = "SwiftPackageManager"
```
What about calling this one `Packages`?

pepibumur commented on Tests/TuistKitTests/Services/Dependencies/DependenciesFetchServiceTests.swift:
```diff
@@ -49,7 +49,11 @@ final class DependenciesFetchServiceTests: TuistUnitTestCase {
                 options: []
             ),
-            swiftPackageManager: nil
+            swiftPackageManager: .init(
```
Nitpick: What about calling this one `packages`?

pepibumur commented on Sources/TuistGraph/Models/SwiftPackageManagerDependencies.swift:
```diff
@@ -10,4 +10,58 @@ public struct SwiftPackageManagerDependencies: Equatable {
+    /// **NOTE** It is a temporary solution until Apple resolves: https://forums.swift.org/t/pitch-package-editor-commands/42224
+    public func stringValue() -> String {
+        """
+        // swift-tools-version:5.3
```
What about using the same Swift version the environment is pointing to. You can use [this method](https://github.com/tuist/tuist/blob/main/Sources/TuistSupport/System/System.swift#L165) to get it.

laxmorek replied:
I have different versions for `Swift` and `Swift tools`.

<img width="514" alt="Zrzut ekranu 2021-03-15 o 21 33 43" src="https://user-images.githubusercontent.com/4774319/111217767-372b1d00-85d6-11eb-9c58-a360ed711fce.png">

It end ups with

<img width="1365" alt="Zrzut ekranu 2021-03-15 o 21 35 35" src="https://user-images.githubusercontent.com/4774319/111217909-62ae0780-85d6-11eb-9258-144ba2be5a67.png">

What do you think about adding new method `Systeming.swiftToolsVersion`?  @pepibumur @fortmarek

pepibumur commented on Sources/TuistGraph/Models/SwiftPackageManagerDependencies.swift:
```diff
@@ -10,4 +10,58 @@ public struct SwiftPackageManagerDependencies: Equatable {
+
+// MARK: - Package.stringValue()
+
+private extension Package {
```
I think this method can be part of `Package.swift`. Moreover, I think `manifestValue` would be more accurate because it's the format of the package when used in a manifest file.

pepibumur commented on Sources/TuistGraph/Models/SwiftPackageManagerDependencies.swift:
```diff
@@ -10,4 +10,58 @@ public struct SwiftPackageManagerDependencies: Equatable {
+
+// MARK: - Requirement.stringValue()
+
+private extension Requirement {
```
Similar comment here. I'd move `stringValue` to `Requirement.swift` _(without extensions)_ and rename it to manifest value.

fortmarek replied:
This was actually my suggestion - in the future, we want to drop this logic altogether when it's possible to use CLI to add package dependencies (which should most likely come with Xcode 13, although, one can never be 100 % sure) and this is a way how to contain this logic. 

If we should ever need it somewhere else, I'd be for moving it to `Requirement.swift`, but for now I believe it makes more sense to keep it here as it's the only place where it's needed.

Alternatively, we could introduce private helper func to `SwiftPackageManagerDependencies` that will be along the lines of `manifestValue(for requirement: Requirement)` instead.

pepibumur commented on Tests/TuistDependenciesTests/SwiftPackageManager/SwiftPackageManagerInteractorTests.swift:
```diff
@@ -1,28 +1,91 @@
 
+    private var fileHandlerMock: MockFileHandler!
+
+    private var temporaryDirectoryPath: AbsolutePath!
```
You don't need to implement this yourself. `TuistTestCase` already provides a [method](https://github.com/tuist/tuist/blob/main/Sources/TuistSupportTesting/TestCase/TuistTestCase.swift#L85) that lazily creates a temporary directory for you.

```swift
func test_whatever() {
  // Given
  let tmpDir = try self.temporaryPath()

}

pepibumur commented on Tests/TuistGraphTests/Models/SwiftPackageManagerDependenciesTests.swift:
```diff
@@ -0,0 +1,74 @@
+@testable import TuistSupportTesting
+
+final class SwiftPackageManagerDependenciesTests: TuistUnitTestCase {
+    func test_stringValue_singleDependency() {
```
If we rename `stringValue` to `manifestValue`, we should rename the tests here too.

pepibumur commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+        try fileHandler.write(packageManifestContent, path: packageManifestPath, atomically: true)
+
+        // log
+        logger.info("Package.swift:", metadata: .subsection)
```
This content is too verbose for the user so I'd rather use `logger.debug`

laxmorek replied:
I changed it in `CarthageInteractor` as well.

pepibumur commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+            try postInstallationActions(pathsProvider: pathsProvider)
+        }
+
+        logger.info("Swift Package Manage dependencies were fetched successfully.", metadata: .subsection)
```
Nitpick: `Packages resolved and fetched successfully`.

laxmorek replied:
I changed it in `CarthageInteractor` as well.

pepibumur commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+            )
+
+            // prepare for installation
+            try prepareForInstallation(pathsProvider: pathsProvider, dependencies: dependencies)
```
When reading the code, it's hard to say what `prepareForInstallation` does. What about coming up with a more specific name. Something along the lines of: `restoreCurrentDependencie`. Maybe @ollieatkinson can come up with a better suggestion here :)

pepibumur commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+            try prepareForInstallation(pathsProvider: pathsProvider, dependencies: dependencies)
+
+            // build command
+            let command = buildCommand(packagePath: temporaryDirectoryPath)
```
I'm not sure we need a method for this if the only thing it does is concatenating a bunch of arguments. I'd rather inline the implementation here.

pepibumur commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+        let command = ["swift", "package", "--package-path", "\(packagePath.pathString)", "resolve"]
+
+        // log
+        logger.info("Command:", metadata: .subsection)
```
You don't need to print this yourself. If the user uses `--verbose`, `system` will automatically print the commands that we are running so this ends up being redundant.

laxmorek replied:
I changed it in CarthageInteractor as well.

pepibumur commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+            let command = buildCommand(packagePath: temporaryDirectoryPath)
+
+            // run `Swift Package Manager`
+            logger.info("SwiftPackageManager:", metadata: .subsection)
```
I think we can remove this one.

laxmorek replied:
I changed it in CarthageInteractor as well.

pepibumur reviewed (commented):
This looks great @laxmorek 👏. Amazing job. One suggestion that I'd make is adding a fixture and a test to test that the resolution of packages works as intended. This can be done in a follow-up PR though.

fortmarek commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+import TuistSupport
+
+enum SwiftPackageManagerInteractorError: FatalError, Equatable {
+    /// Thrown when `Package.resolved` cannont be found in temporary directory after `Swift Package Manager` installation.
```
```suggestion
    /// Thrown when `Package.resolved` cannot be found in temporary directory after `Swift Package Manager` installation.
```

laxmorek replied:
I found similar typo in `CarthageInteractorError`. I fixed it as well.

fortmarek commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+enum SwiftPackageManagerInteractorError: FatalError, Equatable {
+    /// Thrown when `Package.resolved` cannont be found in temporary directory after `Swift Package Manager` installation.
+    case packageResolvedNotFound
+    /// Thrown when `.build` directory cannont be found in temporary directory after `Swift Package Manager` installation.
```
```suggestion
    /// Thrown when `.build` directory cannot be found in temporary directory after `Swift Package Manager` installation.
```

laxmorek replied:
I found similar typo in `CarthageInteractorError`. I fixed it as well.

fortmarek commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+    private let fileHandler: FileHandling
+
+    public init(
+        fileHandler: FileHandling = FileHandler.shared
```
I don't really think we need to be injecting `FileHandler` manually as every `TuistUnitTestCase`should inject `MockFileHandler` by default. You can just use `FileHandler.shared` directly

fortmarek commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+        if fileHandler.exists(toPath) {
+            try fileHandler.replace(toPath, with: fromPath)
+        } else {
+            try fileHandler.copy(from: fromPath, to: toPath)
```
does copy fail if a destination exists? If not, I'd merge `copyFile` and `copyDirectory` methods

fortmarek commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+
+// MARK: - Models
+
+private struct SwiftPackageManagerPathsProvider {
```
I'd either make this internal and hide the implementation behind protocol plus inject this via initializer *or* make these just computed vars which would be fine, too

fortmarek commented on Tests/TuistDependenciesTests/SwiftPackageManager/SwiftPackageManagerInteractorTests.swift:
```diff
@@ -1,28 +1,91 @@
-        subject = SwiftPackageManagerInteractor()
+
+        do {
+            temporaryDirectoryPath = try TemporaryDirectory(removeTreeOnDeinit: true).path
```
as mentioned above, you don't need to initialize this property
if that was not the case, you also wouldn't have to make `do ... catch` dance and just use `setUpWithError` function

fortmarek commented on Tests/TuistDependenciesTests/SwiftPackageManager/SwiftPackageManagerInteractorTests.swift:
```diff
@@ -1,28 +1,91 @@
     }
 
     override func tearDown() {
+        fileHandlerMock = nil
```
there's a mock implementation already created in `TuistUnitTestCase`, so I'd delete this

fortmarek commented on Tests/TuistDependenciesTests/SwiftPackageManager/SwiftPackageManagerInteractorTests.swift:
```diff
@@ -1,28 +1,91 @@
+        XCTAssertTrue(fileHandler.exists(expectedBuildDirectory.appending(components: "artifacts", "foo.txt")))
+        XCTAssertTrue(fileHandler.exists(expectedBuildDirectory.appending(components: "checkouts", "Alamofire", "Info.plist")))
+        XCTAssertTrue(fileHandler.exists(expectedBuildDirectory.appending(components: "repositories", "checkouts-state.json")))
+        XCTAssertTrue(fileHandler.exists(expectedBuildDirectory.appending(components: "repositories", "Alamofire-e8f130fe", "config")))
```
I think we can make the test stronger and more readable:
```swift
XCTAssertEqual(
    fileHandler.contentsOfDirectory(expectedBuildDirectory).sorted(),
   [
       expectedBuildDirectory.appending(component: "manifest.db")),
      .....
   ]
)
```

fortmarek reviewed (commented):
Added some minor comments - great work @laxmorek 👏 

Is there also a fixture one could try this feature with?

laxmorek commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -0,0 +1,190 @@
+    /// Error type.
+    var type: ErrorType {
+        switch self {
+        case .packageResolvedNotFound,
```
I changed it in `CarthageInteractorError` as well.

laxmorek commented:
Hey @pepibumur @fortmarek 

First of all thank you for your reviews. I am grateful for any suggestion and help.

To keep consistency in the codebase every suggestion you left was considered (and applied) for both SPM and (already implemented) Carthage integration. 

Both of you asked for fixture/acceptance test to check. My initial idea was to add it in the next PR but you changed it. I updated `app_with_framework_and_tests_and_dependencies` fixture and `dependencies.feature` acceptance test.

=======

I want to open a few discussion here and Id be glad for your opinion.

### 1 `swift-tools-version`

> What about using the same Swift version the environment is pointing to. You can use this method to get it.

Like I mentioned in the [thread](https://github.com/tuist/tuist/pull/2656#discussion_r594665585) it may `swift` version may be different than `swift-tools-version`. 

Whats more I think `swift-tools-version` should be determined for whole team and shouldn't depend on every single machine. Like I suggested in PR's description I'd add  possibility to define `swift-tools-version` value in `Depedencies.swift` manifest file.

### 2 `SwiftPackageManagerDependencies` vs `Packages`

> What about calling this one Packages?

> Nitpick: What about calling this one packages?

This long name is on propose here. If you look on structure of `Depedencies.swift`

```
Depedencies # Root model
    | - CarthageDependencies # Describes Carthage's depedencies + all configuration around
        | - [Dependency] # Depedencies itself
        | - Set<Platform>   # example of configuration  
    | - SwiftPackageManagerDependencies Describes SPM's depedencies + all configuration around
        | - [Package] # Packages itself
        | - swift-tools-version (?)  # example of configuration  
```

- When I use in code `SwiftPackageManager` or `SwiftPackageManagerDependencies` I mean the description of SPM's packages plus configuration I want to apply
- When I use in code `Package` I mean the package itself

The idea is if you see property named `swiftPackageManager` you should expect the description of packages and options/configurations attached to packages. If you see `package` you should expect the description of package itself (path, version). 

What's more Id like to avoid code like: 

```swift
let packages = packages.packages 
```

I think my differentiation may help with understanding the code. Correct me if you don't agree. 

### 3 `manifestValue`

> I think this method can be part of Package.swift. Moreover, I think manifestValue would be more accurate because it's the format of the package when used in a manifest file.

I used  `manifestValue` I agree it suits better. 

We have here unresolved conversation. Id keep it in private a single file. My intention is to remove it in the future when Apple resolves [this](https://forums.swift.org/t/pitch-package-editor-commands/42224). I can't find any other usage of `manifestValue`.

fortmarek commented:
> Whats more I think swift-tools-version should be determined for whole team and shouldn't depend on every single machine. Like I suggested in PR's description I'd add possibility to define swift-tools-version value in Depedencies.swift manifest file.

Shouldn't `swift-tools-version` be an implementation detail? What would the teams gain by specifying it as they don't know how the manifests we create look like? I'd maybe keep it as you have done it initially @laxmorek and hardcode it as there is no drawback to using older version of package tools (if you have a newer version, you can still compile older version) + we can then be sure that the `Package.swift` we generate will be compilable.

laxmorek commented:
UPDATE: ad.1 

I can see that CI also has a problem because `swift` version doesn't match `swift-tools-version` 

<img width="1462" alt="Zrzut ekranu 2021-03-19 o 09 23 23" src="https://user-images.githubusercontent.com/4774319/111751422-c49b9500-8894-11eb-894d-437fc694e228.png">

fortmarek commented:
> I can see that CI also has a problem because swift version doesn't match swift-tools-version

If we hardcode the tools version, it should be the oldest one that is packaged with the oldest Xcode version we support. It's fine if you specify older one but you can not use swift tools version that is newer than the one available.

laxmorek commented:
> it does not have to do anything with Swift version

it has have since I included [this](https://github.com/tuist/tuist/pull/2656#discussion_r594197613) suggestion.

fortmarek commented on Sources/TuistGraph/Models/SwiftPackageManagerDependencies.swift:
```diff
@@ -10,4 +10,58 @@ public struct SwiftPackageManagerDependencies: Equatable {
+        let package = Package(
+            name: "PackageName",
+            dependencies: [
+                \(packages.map { $0.manifestValue + "," }.joined(separator: "\n        "))
```
wouldn't it be better to use `\t` instead of multiple spaces?

laxmorek replied:
I replaced it but in my opinion tests are less readable now.

fortmarek commented:
> The idea is if you see property named swiftPackageManager you should expect the description of packages and options/configurations attached to packages. If you see package you should expect the description of package itself (path, version).

This distinction feels valid to me. Maybe rather than writing `SwiftPackageManager` we can use a shorter version `SPM` as it it preserves the level of explicitness while making it less verbose.

When trying it out, I have also noticed two things (which are probably unrelated to this PR):
[here](https://github.com/tuist/tuist/pull/2656/files#diff-601113779d98ab55efb9e25a27edaf3ab5e3ec8d9f2c0f3ff4b61792efa4c8a5R47) is a typo

<img width="1407" alt="image" src="https://user-images.githubusercontent.com/9371695/111909963-9e762080-8a5f-11eb-918d-6e6b464e6dbe.png">

I believe we should not be building for macOS platform since the fixture which I ran the command for is iOS only?

laxmorek commented:
### Update

What's is new:
- I restored hardcoded `swift-tools-version:5.3`
- I added cleaning unnecessary dependencies (check `DependenciesController.swift` for details) 
- I removed `macOS` from fixture as Marek had suggested

> we can use a shorter version SPM as it it preserves the level of explicitness while making it less verbose.

Personally I prefer long name. We had this discussion [here](https://github.com/tuist/tuist/pull/2227). 

Anyway, shorter version works for me too so if you strongly recommend it I can replace it.

fortmarek reviewed (approved):
> Anyway, shorter version works for me too so if you strongly recommend it I can replace it.

sure, I don't mind it as-is as well 👍 I don't have anything to add, the PR seems to be in good state from my point of view, anyways. Let's wait for @pepibumur to have a second look at this 🙂

pepibumur reviewed (approved):
Hey @laxmorek. Awesome job! I think we can merge when CI is green. I pushed a couple of typo fixes.
